tests/federation_acl_test.go

```diff
@@ -0,0 +1,142 @@
+
+	// 3. Join this room from 2nd server
+	bob.JoinRoom(t, roomID, []string{"hs1"})
+	aliceSince = alice.MustSyncUntil(t, client.SyncReq{Since: aliceSince}, client.SyncJoinedTo(bob.UserID, roomID))
+	bobSince := bob.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(bob.UserID, roomID))
+
+	// create a different room used for a sentinel event
+	sentinelRoom := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+	aliceSince = alice.MustSyncUntil(t, client.SyncReq{Since: aliceSince}, client.SyncJoinedTo(alice.UserID, sentinelRoom))
+	bob.JoinRoom(t, sentinelRoom, []string{"hs1"})
+	charlie.JoinRoom(t, sentinelRoom, []string{"hs1"})
+	aliceSince = alice.MustSyncUntil(t, client.SyncReq{Since: aliceSince},
+		client.SyncJoinedTo(bob.UserID, sentinelRoom),
+		client.SyncJoinedTo(charlie.UserID, sentinelRoom),
+	)
+
+	// 4. Add deny rule, to block 2nd server from participating
+	stateKey := ""
+	eventID := alice.SendEventSynced(t, roomID, b.Event{
+		Type:     "m.room.server_acl",
+		Sender:   alice.UserID,
+		StateKey: &stateKey,
+		Content: map[string]interface{}{
+			"allow":             []string{"*"},
+			"allow_ip_literals": true,
+			"deny":              []string{"hs2"},
+		},
+	})
+	// wait for the ACL to show up on hs2
+	bob.MustSyncUntil(t, client.SyncReq{Since: bobSince}, client.SyncTimelineHasEventID(roomID, eventID))
+
+	// 5. Join from 3rd (dendrite) server.
+	charlie.JoinRoom(t, roomID, []string{"hs1"})
+	aliceSince = alice.MustSyncUntil(t, client.SyncReq{Since: aliceSince}, client.SyncJoinedTo(charlie.UserID, roomID))
+	charlieSince := charlie.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(charlie.UserID, roomID))
+
+	// 6. Any events sent by 2nd server will not appear for users from 1st server, but will be visible for users from 3rd server
+	eventID = bob.SendEventSynced(t, roomID, b.Event{
+		Type:   "m.room.message",
+		Sender: bob.UserID,
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "I should be blocked",
+		},
+	})
+
+	// sentinel event in room2
+	sentinelEventID := bob.SendEventSynced(t, sentinelRoom, b.Event{
+		Type:   "m.room.message",
+		Sender: bob.UserID,
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "I should be visible",
+		},
+	})
+	// wait for the sentinel event to come down sync
+	alice.MustSyncUntil(t, client.SyncReq{Since: aliceSince}, client.SyncTimelineHasEventID(sentinelRoom, sentinelEventID))
+	charlie.MustSyncUntil(t, client.SyncReq{Since: charlieSince}, client.SyncTimelineHasEventID(sentinelRoom, sentinelEventID))
+
+	// Verify with alice and charlie that we never received eventID
+	for _, user := range []*client.CSAPI{alice, charlie} {
+		syncResp, _ := user.MustSync(t, client.SyncReq{})
+
+		// we don't expect eventID (blocked) to be in the sync response
+		events := syncResp.Get(fmt.Sprintf("rooms.join.%s.timeline.events", client.GjsonEscape(roomID))).Array()
```

Comment:
What guarantees that this initial sync response includes the sentinel event? (Synapse caches initial sync responses, though this is [sneakily disabled in complement](https://github.com/matrix-org/synapse/blob/22cc93afe38d34c859d8863a99996e7e72ca1733/docker/complement/conf/workers-shared-extra.yaml.j2#L115-L117).)